Feat/sparse kernel

felixmusil opened the pull request:
This PR brings a few new features to the code to enable sparse GPR training with energy and forces and predictions of  energy and forces using rascal and Spherical Invariants:
1. a Sparse Kernel class that mimics GAP
2. a Pseudo points class tailored for models 
3. c++ and python class serialization
4. python krr model with GAP trainer

An example of building an interatomic potential for a small molecule is in MLIP_example.ipynb

1) The class is meant to compute kernel matrices that are used in the sparse GPR methods like Subset of Regressors. It is pretty straight forward and most details are in the doc of the class and maybe a more comprehensive descriptions in the docstring of the train_gap_model() function.

2) Pseudo points are key to many sparse GPR models, they basically corresponds to interpolation points in the regression methods. It aims at reducing the cost of training and prediction from N^3/N^2 to ~MN^2/M^2 where N and M are respectively the number of training and pseudo points. We typically choose pseudo points from the training set though.

3) after training a model we need to save it whether it exists in python or c++. At the moment I only implemented a python model so it need to be saved/loaded through python. I choose python dictionaries as the serialization format as it can then be written to disk in most cross platform format like JSON, msgpack... so potentially a model saved from python could be read and ran from c++
To do that one needs to serialize python objects to dictionaries (done thought the BaseIO class and some modifications of the class that should be saved). 
And also the underlying c++ objects which is done in a 2 step procedure:  (a) convert the c++ class to the json class (see the end of the SparseKernel file for instance) in c++ (b) then convert the json c++ object to a python dictionary in the to_dict/from_dict binding functions.
Note that the main object that needs serialization is the PseudoPoints class as it holds a lot of the training information stored in a sparse format when dealing with SphericalInvariants features.

4) the GAP training is a particular procedure defined in QUIP. It uses sparse kernels and a particular kernel to train with energy/forces and predicts energy/forces. I am replicating this model because it works quite well and 'many' people are familiar with it.

Note that most of the diff comes from the update of the json.hpp to a newer version.

mastricker commented:
If I should review on this, I need a little bit more description on the what and how and so on. Maybe in the next meeting?

I can definitely look over the code, but I am probably not the right one for checking functionality/sensibility.

Pseudo-points? What do you need that for? I am not familiar with the term.
FF -> force field? That is probably clear - meaning you also train on forces?
GAP trainer -> A training environment for Gaussian Approximation Potentials? What does that all entail?
Serialization for i/o c++/python or to for storage of the model?

Sorry, I need more info....or someone else does the logic check and I just look at code as code.

Luthaf commented:
> a Sparse Kernel class that mimics GAP
> a Pseudo points class tailored for models
> c++ and python class serialization
> python krr model with GAP trainer

Each one of these points is a big change that would warrant discussion by itself =). It would be nicer to split this PR in four separated set of changes that can be discussed and reviewed independently.

> a Sparse Kernel class that mimics GAP

I would like to see more justification about why this is needed inside rascal, instead of using pre-existing code (sklearn or other), as we started discussing the other day 😃

felixmusil commented:
> Each one of these points is a big change that would warrant discussion by itself =). It would be nicer to split this PR in four separated set of changes that can be discussed and reviewed independently.

The changes are interlinked so I would say that it would take me more time to split them properly than to walk through all of them in one PR.

> I would like to see more justification about why this is needed inside rascal, instead of using pre-existing code (sklearn or other), as we started discussing the other day smiley

I think looking at the implementation of the KRR class and comparing it to the sklearn one is quite self explanatory. We can do that together if you prefer.

Luthaf commented:
> I think looking at the implementation of the KRR class and comparing it to the sklearn one is quite self explanatory. We can do that together if you prefer.

I think I am able to read through the code by myself. But this does not answer my question: I think you are implying this implementation is simpler than others, but I don't see why this is enough to decide in favor of it or against it. 

What about generality (different kernels, dealing with normalization and centering); the presence of other methods with the same interface (allowing one to test multiple methods easily); and performance (more complex code might be better optimized)?

> The changes are interlinked so I would say that it would take me more time to split them properly than to walk through all of them in one PR.

Well, this saves you a bit of time, but make it harder for reviewers to understand the motivation for each of these changes, think whether the solution proposed here is a good one, and understand the code. So you may be saving some time, at the cost of others loosing some.

mastricker commented:
In reference to Guillaume's points, I made an entry in the agenda for the next meeting. I think this discussion is bigger and we probably should speak a little bit about the development process.

felixmusil commented:
> I think you are implying this implementation is simpler than others

I am saying the opposite actually since our needs are more 'feature rich' than what is provided out of the box in sklearn for instance. Also note that regression models in sklearn are based on a `fit`/`predict` functionality which we decided to not follow (already discussed in several meetings) in favor of a collection of external functions/class dedicated to training and models that only predict.

> What about generality (different kernels, dealing with normalization and centering); the presence of other methods with the same interface (allowing one to test multiple methods easily); and performance (more complex code might be better optimized)?

Could you be a bit more specific ? what different kernel ? what other methods ? what other code ?

> this saves you a bit of time

Unfortunately I am not familiar with git commit rewriting so I expect such split of the PR to take me at least a day if not much more since many commits in addressing several points. I don't feel that this is 'just' a bit of time.

Luthaf commented:
> Could you be a bit more specific ? what different kernel ? what other methods ? what other code ?

Yes, sorry. I was thinking of kernels other than the cosine kernel (gaussian, Mattern, Fisher, etc.). Other methods would include non KRR methods still using kernels for example SVM, Kernel Independent Component Analysis, or the KPCovR thing we are working on with Rose & Ben. Other codes refer to other project than sklearn that other people can use with such methods: pytorch/sktorch, etc.

> I am saying the opposite actually since our needs are more 'feature rich' 

OK, that was not clear to me. What in particular is needed?

felixmusil commented:
> I was thinking of kernels other than the cosine kernel (gaussian, Mattern, Fisher, etc.).

these kernels correspond to a function of the distance so they can be built by first computing the cosine kernel, transform it into the l_2 distance and then apply the function element wise so I would say that we are still quite general since the 2 last steps can be done in either c++ or python without big differences.

> Other methods would include non KRR methods still using kernels for example SVM, Kernel Independent Component Analysis, or the KPCovR thing we are working on with Rose & Ben.

SVM is tricky enough to implement that it is worth using something that is already out there clearly, and the other two, I am not sure but I guess it should be pretty straight forward to adapt their python implementation so that they can rely on rascal to compute the kernel.  

> Other codes refer to other project than sklearn that other people can use with such methods: pytorch/sktorch, etc.

to use rascal along with, let's say, pytorch will require a dedicated interface that does not exist and that will probably be valid only for pytorch so why is it relevant to this PR ?

> What in particular is needed?

3 main things: the model is built trained, it has a central atom type dependent offset (isolated atom energies for instance) and it can predicts the gradient of the property w.r.t to the atomic positions.

mastricker commented on tests/test_pseudo_points.cc:
```diff
@@ -0,0 +1,206 @@
+      int ii{0};
+      for (auto center : manager) {
+        (void)center;  // to avoid compiler warning
+        double roll{dice(gen)};
```
Is the rationale here that the _pseudo points_ are chosen randomly, because it is just a test? I am not sure I understand this test - the reference to which it is compared `sparse_points_b` are chosen how exactly from the representation?

felixmusil replied:
this tests picks randomly (for convinience) some pseudo points from a set of structures and serialize to json and deserialize the json and compare this new object with the original one.  I updated the test description, tell me if it makes more sense.

mastricker commented on tests/test_pseudo_points.cc:
```diff
@@ -0,0 +1,206 @@
+      PseudoPointsFixture<CalculatorSphericalInvariants, ManagerCollection_t,
+                          PseudoPointsBlockSparse>>;
+  /**
+   * Tests if the wrong target_type is catched correctly.
```
```suggestion
   * Tests if the wrong target_type is caught correctly.
```

mastricker commented on tests/test_pseudo_points.cc:
```diff
@@ -0,0 +1,206 @@
+    double cutoff{3.};
+    json hypers{{"max_radial", 1},
+                {"max_angular", 1},
+                {"compute_gradients", true},
```
Are gradients needed to check the pseudo points?
Gradients themselves are check somewhere else, right?
Maybe here is a possibility to have a smaller test.

felixmusil replied:
right nice catch, its en unfortunate copy/paste.

mastricker commented on src/rascal/models/pseudo_points.hh:
```diff
@@ -0,0 +1,419 @@
+  /**
+   * Set of pseudo points associated with a representation stored in BlockSparse
+   * format, e.g. SphericalInvariants. The number of pseudo points is often
+   * refered as $M$ and note that they might be the representation of actual
```
```suggestion
   * referred as $M$ and note that they might be the representation of actual
```

mastricker commented on src/rascal/models/pseudo_points.hh:
```diff
@@ -0,0 +1,419 @@
+    //! counts number of sparse points for each [sp]
+    Counters_t counters{};
+    //! size of one feature block in [sp][key]
+    size_t inner_size{0};
```
Are those not dependant on the species? Are they the same for all blocks. I think I don't understand.

felixmusil replied:
Yes, in the case of the power spectrum inner_size will be n_max**2*(l_max+1) for all species. Of course one could always think of having different expansions parameters for different central atom type but at the moment such strategy is not possible so not suported. Simple fix would be to have a map instead of just one size_t.

mastricker replied:
I.e. for now this is reasonable because we do not expand differently for different species?

felixmusil replied:
yes and the machinery in spherical_invariants does not allow for that in an integrated manner.

mastricker commented on src/rascal/models/pseudo_points.hh:
```diff
@@ -0,0 +1,419 @@
+
+    PseudoPointsBlockSparse() {
+      // there is less than 130 elemtents
+      for (int sp{1}; sp < 130; ++sp) {
```
Do I understand correctly that `PseudoPoints` is always `PseudoPointsBlockSparse`.
Is there any other way to look at this? Could there possibly be a non-sparse `PseudoPoints` class?
If so, should there be a base class and this a specialization of it?

felixmusil replied:
`PseudoPointsBlockSparse` is only compatible with `PseudoPointsBlockSparse` and yes there should be a `PseudoPointsDense` implementation but it is outside of this PR. I thought of making these with a base class but it felt really pointless here since it would have no implementation functions or virtual ones since all member functions are templated. 
Would you prefer to have an empty base class anyway ?

mastricker replied:
We should probably leave it at that for now before we did not talk about the whole PseudoPoints business. Maybe some stuff will clear up.

Luthaf commented:
> Does anyone know an easy way to separate out code to just to a partial PR?

I am not sure what you mean by that, but one can `squash` all commits, soft-`reset` to HEAD^, and then pick related pieces of code using `git add -p` to create a single commit for each functionality, then create branches & PR with a single commit at each time. It is a bit of work, especially if the initial code was not neatly decoupled, meaning that it is something to keep in mind when working. If one find him/herself working on something unrelated while on `feat/my_maxi_feature`, they should create a new branch `feat/a_required_feature`, work on this separate branch and rebase `feat/my_maxi_feature` on top of `feat/a_required_feature` regularly.

felixmusil commented:
Thank you very much @mastricker  for opening the review process.

> But what I also see design-wise, it seems more like a `BlockSparseProperty`. E.g. if a calculator's output is stored in a property a choice of `PseudoPoints` from those features are a different
arrangement/subset of that property. 

I chose to implement `PseudoPoints` in the model folder for 2 reasons. `PseudoPoints` are only needed because there are `SparseKernel`, such abstraction does not make sense in any other context. Moreover, even if it seems quite similar to Properties they are essentially different since `PseudoPoints` are not related to atomic structures, i.e. they are basis points for the sparse kernel model. It is only coincidental that we generate these basis points using features extracted from atomic structures. Of course we should discuss this further next week.

> I know it is easy to pick on these things, but here we go: In general, the
python code looks not nice. Inconsistent formatting, too long lines, missing
whitespaces, also the docstrings - what happened to the autopepper? This
seems to be getting out of hand on the python side.

Hum... I ran several times `pretty-python` and did not really check whether it did its job properly. I will look into it.

> Do we really need 24.000 new lines of reference data input?
molecule_conformers_dftb.xyz??

Its the data needed for the IPML_example.ipynb to run smoothly. I could crop it maybe ? since few people have enough RAM to build the interatomic potential with more than 2k structures. The point is that we don't have a reference dataset with forces and energies to train a model at the moment so some additional inputs are needed.

> Units: In the calculators, the units are not checked.

I am not sure of the state of the unit checker and how it should be used so I made sure that units are there and the user should be careful at the moment. When it becomes clearer this check will be added to the calculator_base class so that its easy to make the check in each calculator.

> Other questions: I saw that the keyword override was added to some of the
functions - was that necessary only now because different calculators have
different arguments? Or why was virtual not enough? Was some indirection wrong?

some compiler on the CI started to complain about missing `override` so I added it.

> And at last: This all needs proper written documentation. If not for me, for
anyone else who wants to adapt this functionality. This new functionality is
propably very useful in the near future. I personally need
drawings/figures/schematics for this. Maybe this is now the time to define the
(mentioned above) concept of a Model.

You are totally right but I am not sure about how this documentation should be framed in terms of content and objectives so I would like to first sort this out (maybe next meeting ?).

felixmusil commented on bindings/bind_include.hh:
```diff
@@ -54,9 +55,59 @@ PYBIND11_MAKE_OPAQUE(std::vector<rascal::AtomicStructure<3>>);
 namespace py = pybind11;
 
-namespace rascal {
+//! Simplistic but robust implicit conversion of py::dict to/from nlohmann::json
```
right it's only half implicit. What I mean is that you can do:
```
json j = py::dict("one"_a=1, "b"_a="fgssdf");
py::dict d = j.get<py::dict>();
```
and these 2 functions are taking care of doing that. Should I just put this example in the comment ?

mastricker replied:
Yes, I like that.

felixmusil commented on bindings/bind_include.hh:
```diff
@@ -54,9 +55,59 @@ PYBIND11_MAKE_OPAQUE(std::vector<rascal::AtomicStructure<3>>);
+        json j = self;  // implicit conversion to nlohmann::json
+        std::string str = j.dump(2);
+        std::string representation_name{internal::type_name<Object>()};
+        std::string sep{" | Parameters: "};
```
This function is used by all bindings to bind the to_dict / from_dict / __str___ functions so it is already quite global I would say. Also this is here only for a 'nicer' string representation of the objects, i.e. `print(my_binded_object)` returns something more meaningful that the class name, so its just to make some binded object a bit more user friendly. I feel that its a bit overkill to test such non essential features (nobody should rely on this output in code). We probably should discuss this next meeting.

mastricker replied:
Hmm,ok. I see your point. Taken.

felixmusil commented on bindings/bind_py_models.cc:
```diff
@@ -58,6 +59,59 @@ namespace rascal {
+
+  //! register the population mechanism of the pseudo points class
+  template <class ManagerCollection, class Calculator, class PseudoPoints>
+  void bind_pseudo_points_push_back(py::class_<PseudoPoints> & pseudo_points) {
```
I chose `extend` to follow python's list terminology: `append` would be appropriate if you were adding just one element but since we give a structure and potentially several elements are added then you are extending the list. 
Would you rather keep the same names in python and c++ for binded functions ?

mastricker replied:
Point taken.

felixmusil commented on bindings/bind_py_models.cc:
```diff
@@ -58,6 +59,59 @@ namespace rascal {
+        py::overload_cast<const Calculator &, const StructureManagers &,
+                          const PseudoPoints &>(
+            &SparseKernel::template compute<Calculator, StructureManagers,
+                                            PseudoPoints>),
```
so the first compute function is there to build elements of the type: `K(A,T)` where A is an atomic structure and T is a pseudo point. The second is for elements of the type `K(T,T)`. So to me it is the same functionality hence the overload. Do you still think its different ?

mastricker replied:
Understood. Ok.

felixmusil commented on bindings/bind_py_models.cc:
```diff
@@ -67,7 +121,7 @@ namespace rascal {
    *
    */
-  void add_kernels(py::module & mod, py::module & m_internal) {
+  void add_models(py::module & mod, py::module & m_internal) {
```
I renamed the function because it effectivelly binds more than just kernel function and to me it's fair to bundle all these functionalities under the model umbrella but its true that we have not discuss this in details already.

mastricker replied:
I see. I made an entry in the meeting minutes.

felixmusil commented on bindings/bind_py_representation_calculator.cc:
```diff
@@ -40,15 +40,12 @@ namespace rascal {
+    representation.def(py::init([](const py::dict & hyper) {
       // convert to json
-      json hypers = json::parse(hyper_str);
+      json hypers = hyper;
```
There is a 'bad' interaction with the json construction and initialiazer list. Since an initialiazer list is a list the resulting json object is always a json::array of what you put in the braces but here we want to convert the py::dict to a json::object.

felixmusil commented on bindings/rascal/models/kernels.py:
```diff
@@ -24,23 +27,33 @@ class Kernel(object):
+        Y : AtomList, ManagerCollection or PseudoPoints* (C++ class).
+
+        grad : tuple specifying if the kernel should be computed using gradient
+               of the representation w.r.t the atomic positions.
```
There are 2 inputs to the kernel function so what I mean is that you can choose between taking the derivative with respect to the 1st and/or the second argument of the kernel function. Note only the gradient with the 1st is implemented :)
Is it clearer now ?

felixmusil commented on bindings/rascal/models/kernels.py:
```diff
@@ -52,19 +65,67 @@ def __init__(self, representation, name='Cosine', target_type='Structure',
             if not(zeta > 0 and zeta % 1 == 0):
```
I would rather keep the posibility for getting a 2.0 but I am needing an int is also fine.

mastricker replied:
I am not sure I understand what you are saying. So you agree or disagree with the suggestion?

felixmusil replied:
changed

felixmusil commented on bindings/rascal/models/kernels.py:
```diff
@@ -52,19 +65,67 @@ def __init__(self, representation, name='Cosine', target_type='Structure',
+        elif (name == 'GAP' and 'zeta' in kwargs):
+            # should be positive integer
+            zeta = kwargs['zeta']
+            if not(zeta > 0 and zeta % 1 == 0):
```
same as above

felixmusil commented on bindings/rascal/models/kernels.py:
```diff
@@ -52,19 +65,67 @@ def __init__(self, representation, name='Cosine', target_type='Structure',
 
-    def __call__(self, X, Y=None):
+    def _get_data(self):
+        return dict()
```
yes there is no data to return here so the dict is empty but it still has to be a dict to be compatible with the to_dict() function in BaseIO. The output of to_dict() is a nested dictionary into which entries are either dictionaries or lists. It could also allow for `None` I guess but I don't see why it should. Is it clearer ?

mastricker replied:
Yes - so to rephrase, it is because of the definition of the base class.

felixmusil replied:
you are right.

felixmusil commented on bindings/rascal/models/kernels.py:
```diff
@@ -52,19 +65,67 @@ def __init__(self, representation, name='Cosine', target_type='Structure',
+            elif self.kernel_type == 'Sparse':
+                if isinstance(X, PseudoPoints):
+                    X = X._pseudo_points
+                # compute the KMM matrix
```
This is documented in the train_gap_model function where it is really relevant. I added some explanation and a reference to where the full info is.

felixmusil commented on bindings/rascal/models/krr.py:
```diff
@@ -0,0 +1,203 @@
+        reference samples used for the training
+
+    self_contributions : dictionary
+        map atomic number to the property baseline, e.g. isolated atoms
```
Depending on the property it can refer to different things. Isn't the example clear enough ?

felixmusil commented on bindings/rascal/models/krr.py:
```diff
@@ -0,0 +1,203 @@
+        if compute_gradients is False:
+            return Y0+np.dot(KNM, self.weights).reshape((-1))
+        else:
+            return np.dot(KNM, self.weights).reshape((-1,3))
```
when compute_gradients == True it predicts `-forces` so the 3 is for the directions.

felixmusil commented on bindings/rascal/models/krr.py:
```diff
@@ -0,0 +1,203 @@
+        return self.weights
+
+    def set_calculator(self):
+        pass
```
removed

felixmusil commented on bindings/rascal/models/krr.py:
```diff
@@ -0,0 +1,203 @@
+        return init_params
+
+    def _set_data(self, data):
+        pass
```
it a requirement from BaseIO but there is nothing to do in this case

mastricker replied:
Maybe the requirements from `BaseIO` have to be rethought?

felixmusil replied:
How so ?

felixmusil commented on bindings/rascal/models/krr.py:
```diff
@@ -0,0 +1,203 @@
+        build KNMp.
+    managers : AtomsList
+        Training structures with features (and gradients)
+    KNMp : np.array
```
It does not refer to anything in particular, I just need to have 2 KNM matrices in the function but only one proper name. What would you recommand ?

mastricker replied:
Ok, that is a tricky one. Now that I think of it, maybe something like `KNM2` - the letter suggest some meaning, at least to me.

felixmusil commented on bindings/rascal/models/krr.py:
```diff
@@ -0,0 +1,203 @@
+        KNM = np.vstack([KNM, KNM_down])
+        when training with derivatives.
+    X_pseudo : PseudoPoints
+        reference samples to use in the model
```
I did not think 'reference' would be confusing here. I mean that `X_pseudo` contains the samples that will be used by the model as basis points. Is is better now ?

mastricker replied:
This seems to play into the point @max-veit raised about the naming of the concept. I agree with him. It is also noted it in the meeting minutes.

felixmusil commented on bindings/rascal/models/krr.py:
```diff
@@ -0,0 +1,203 @@
+    y_train : np.array
+        reference property
+    self_contributions : dictionary
+        map atomic number to the property baseline, e.g. isolated atoms
```
self_contributions is a map that link atomic number to property baseline. What is the problem ?

mastricker replied:
What I meant with normalization is: the values in `self_contributions` serve to normalize (or maybe _calibrate_ is the better word) to the per-atom reference energy?
Property baseline also is a term that I find confusing - maybe reference energy?

felixmusil replied:
the property does not have to be energy, e.g. in NMR chemical shifts you have also some baseline shielding for each atomic type. 

Calibrate might be a better term but it also seems off for the case of transforming total energies (output of a DFT code for instance) to formation energies. 

I reformulated the comments. Tell me if it is clearer now ?

felixmusil commented on bindings/rascal/models/krr.py:
```diff
@@ -0,0 +1,203 @@
+
+    Returns
+    -------
+    KRR
```
it's the type of the returned object. I think that is expected in the the numpy docstring format. If it where to return a list it would be written list there.

mastricker replied:
Ok, I see the issue.

felixmusil commented on bindings/rascal/models/krr.py:
```diff
@@ -0,0 +1,203 @@
+    Y = Y - Y0
+    delta = np.std(Y)
+    # lambdas[0] is provided per atom
+    KNM[:n_centers] /=  lambdas[0] / delta * np.sqrt(Natoms)[:,None]
```
if the training happens with forces yes it has `n_centers+n_forces*3` entries. It is explained in the article cited in the docstring, should I also go into the details in the code too ?

mastricker replied:
Maybe not in all the details, but a comment on why subindexing is used here.
At least I understand the KNM matrix now, I guess.

felixmusil replied:
I added some comments

felixmusil commented on bindings/rascal/representations/spherical_covariants.py:
```diff
@@ -160,40 +159,40 @@ def get_num_coefficients(self):
-                                   self.hypers['lam'])**2 / 2.0) -
-                            (self.hypers['max_angular'] -
+                         (1.0 + np.floor((self.hypers['lam'] - 1) / 2.0))**2
+                         - np.floor((self.hypers['max_angular'] + 1
```
I am not super clear on this part, @andreagrisafi or @mjwillatt wrote this bit.

mastricker replied:
I am not questioning the numbers or intend. What I am trying to say: can we refactor this piece of code in terms of the individual contributions (guess that is what @max-veit means with n, lambda, mu, ...) and then make it look like

`n_entries = n + n' + lambda + mu + some_other_value`

To be clearer on what is being done here.

felixmusil replied:
I understand but what I am saying is that I don't know where these contributions come from and it is beside the scope of this PR.

mastricker replied:
Point taken. I'll make an issue out of this and assign the mentioned two people.

felixmusil commented on bindings/rascal/representations/spherical_expansion.py:
```diff
@@ -221,13 +213,13 @@ def transform(self, frames):
+    def get_num_coefficients(self, n_species=1):
         """Return the number of coefficients in the spherical expansion
 
         (this is the descriptor size per atomic centre)
```
it does hence the `n_species` parameter so that the user can get the corresponding number of coefficient. Is it not clear ?

mastricker replied:
No, I am not following. Are they the same for each atomic species? And looking through the code it seems to me that no other value than `n_species=1` is used?

felixmusil commented on bindings/rascal/utils/cur.py:
```diff
@@ -0,0 +1,157 @@
+import numpy as np
+from scipy.sparse.linalg import  svds
+
+def do_CUR(X, Nsel, act_on='sample', is_deterministic=False,seed=10,verbose=True):
```
> choose the error

what do you mean ?

mastricker replied:
My understanding of CUR is that you have two options to approximate the matrix:
1. You choose the number of columns you want to retain, which results in an error (depending on the quality of your descriptor)
2. You choose the error for the approximation and this results in a minimum number of columns which are chosen/added until you approximate the initial matrix accordingly

My question is therefore: Do we also need the option to not only choose the number of columns and live with an error, but also be possible to have the second case, where you specify the error and live with the corresponding number of columns.

Is it clearer?

felixmusil replied:
I see right. I am not sure there is a direct link between reconstruction error and model accuracy except for extreme cases. We typically choose the number of coloumns/rows. 
Would you have a use for a reconstruction error threshold setting ?

felixmusil commented on bindings/rascal/utils/cur.py:
```diff
@@ -0,0 +1,157 @@
+
+
+class CURFilter(BaseIO):
+    """CUR decomposition to select samples or features in a given feature matrix. Wrapper around the do_CUR function for convenience.
```
The functionality is quite general so I prefer to keep the general part in a separate function to make a clear distinction between the selection algorithm and what needs to be done to use it within rascal. An alternative I see would be a static member function but it seems a bit overkill ?

mastricker replied:
Agreed.

max-veit commented on src/rascal/models/sparse_kernels.hh:
```diff
@@ -0,0 +1,498 @@
+        return this->compute_helper<Property_t, TargetType::Atom>(
+            representation_name, managers, sparse_points);
+      default:
+        throw std::logic_error("The combination of parameter is not handled.");
```
This error message keeps on sneaking in, despite it being unreachable code. `TargetType` is an enum; it will only ever go to the cases enumerated in `TargetType`, all of which are covered in this switch statement.

Is the purpose of this message to catch programmer error, where we add a case to the enum but forget to add it to the switch statement? In which case, shouldn't the error message be more informative, giving the value of the enum that wasn't handled? And in any case, aren't there static analysis tools capable of finding such predictable mistakes?

felixmusil replied:
having a default in a switch over an `enum` disables the compiler warnings about missing enum values list of `case:`. Should I remove the default here ?

mastricker replied:
I like the idea of having the logic error to tell you when you forgot to add it. I would leave it.

felixmusil replied:
made modification of the text

max-veit commented on src/rascal/models/sparse_kernels.hh:
```diff
@@ -0,0 +1,498 @@
+    /*
+     * The root compute kernel function. It computes the kernel the
+     * representation gradients of a set of structures and the set of pseudo
+     * points.
```
I think it's a "between", actually

felixmusil commented on bindings/rascal/utils/cur.py:
```diff
@@ -0,0 +1,157 @@
+            # get various info from the structures about the center atom species and indexing
+            sps = list(self.Nselect.keys())
+            types = []
+            strides_by_sp = {sp:[0] for sp in sps}
```
the sorting happens on the c++ side but otherwise that's it.

felixmusil commented on bindings/rascal/utils/io.py:
```diff
@@ -0,0 +1,343 @@
+import json
+from copy import deepcopy
+
+BETA_VERSION = '0.1'
```
There could be several of these version and when you read a file you look at what version it uses and will use the proper reader. I am not sure to follow where the debt is ? 
 I will make functions to fetch this kind of data.

felixmusil replied:
I will add such functions. 

I don't see the future debt here. The logic is the following: there might be several versions of writing to a file over time so several reader are around and when reading the reader will choose which one to use.

felixmusil replied:
I added a `rascal.utils.get_current_io_version` and `rascal.utils.get_supported_io_versions` functions. Is it good now ?

mastricker replied:
I'll have a look over the new files after I'm done with the comments. But it seems to be a better place there.

felixmusil commented on src/rascal/models/pseudo_points.hh:
```diff
@@ -0,0 +1,419 @@
+
+namespace rascal {
+
+  namespace internal {}  // namespace internal
```
I am not sure what version of the file this is. Defenetly not the one I have localy. There is a lot of content in this namespace.

felixmusil commented on src/rascal/models/pseudo_points.hh:
```diff
@@ -0,0 +1,419 @@
+      return species;
+    }
+    //! dot product with itself to build the K_{MM} kernel matrix
+    math::Matrix_t self_dot(const int & sp) const {
```
its the only non templated function that has more than 5 lines of implementation. I would rather keep it close to the other 'dot' functions.

mastricker replied:
maybe start adding an "_impl.hh" file...

felixmusil commented on src/rascal/models/pseudo_points.hh:
```diff
@@ -0,0 +1,419 @@
+     * sp with the representation associated with a center.
+     * @return column vector Mx1
+     */
+    template <class V>
```
yes, `InternallySortedKeyMap` has the template parameter of a map: Key and Value.

felixmusil commented on src/rascal/models/sparse_kernels.hh:
```diff
@@ -0,0 +1,498 @@
+        for (const int & sp : pseudo_points.species()) {
+          // only the pseudo points of the same species contribute
+          auto block_size = pseudo_points.size_by_species(sp);
+          KMM.block(start, start, block_size, block_size) =
```
This matrix is block diagonal so here you compute the blocks. I am not sure what is there to add.

mastricker replied:
Ok, I got confused about the `.block()` - I remembered it as `.block(start1, end1, start2, end2)`, where in fact it is `.block(start1, start2, end1, end2)`.

felixmusil commented on src/rascal/models/sparse_kernels.hh:
```diff
@@ -0,0 +1,498 @@
+            // compute the gradient of the kernel w.r.t the representation dk/dX
+            // without the pseudo point factor
+            math::Matrix_t rep(manager->size(), n_pseudo_points);
+            for (auto center : manager) {
```
I am not sure about what you mean ; all the functions here are templated so they can't be put in a `.cc`, right ?

mastricker replied:
...again `_impl.hh` - Is in the meeting minutes.

felixmusil commented on src/rascal/models/sparse_kernels.hh:
```diff
@@ -0,0 +1,498 @@
+      } else {
+        throw std::logic_error("Requested SparseKernel \'" + kernel_type_str +
+                               "\' has not been implemented.  Must be one of" +
+                               ": \'GAP\'.");
```
I am not sure to follow. The input `kernel_type_str` is a string and you want to see if is matches an `enum` ? I don't know how to compare enums and strings.

mastricker replied:
My bad, I see my wrong logic here now.

felixmusil commented on src/rascal/models/sparse_kernels.hh:
```diff
@@ -0,0 +1,498 @@
+      using internal::SparseKernelType;
+
+      if (this->kernel_type == SparseKernelType::GAP) {
+        auto kernel =
```
you need a functional object right ? so it has to be downcasted from the base which is not functional.

mastricker replied:
I see my error here. On another note: It is already a pointer, right? So no `auto &` necessary?

felixmusil replied:
it is held in a shared_ptr so here it creates a new shared_ptr to the object so no reference is actually possible.

felixmusil commented on src/rascal/representations/calculator_spherical_invariants.hh:
```diff
@@ -172,16 +172,25 @@ namespace rascal {
         : CalculatorBase{}, rep_expansion{hyper} {
       this->set_default_prefix("spherical_invariants_");
       this->set_hyperparameters(hyper);
+      this->hypers = hyper;
```
yes

felixmusil commented on src/rascal/structure_managers/property_block_sparse.hh:
```diff
@@ -310,6 +310,18 @@ namespace rascal {
+        return this->flattened(skey);
+      }
+
+      Eigen::Map<const math::Vector_t> flattened(const SortedKey_t & skey) {
```
it gives a flattened view of the object which is why I did not called it `flatten`. If you think the distinction is irrelevant I will rename it.

mastricker replied:
I would argue to call it `flatten` - to be closer to numpy.

max-veit replied:
numpy terminology would be `flat`, in fact

felixmusil replied:
it's not quite the same as `flat` but it is more similar I agree. I renamed to flat

mastricker replied:
Isn't it the same as [here](https://numpy.org/doc/1.18/reference/generated/numpy.ndarray.flatten.html)

max-veit replied:
"Return a _copy_ of the array flattened into one dimension". Copy, not view, so no.

felixmusil commented on src/rascal/utils/json_io.hh:
```diff
@@ -32,8 +32,17 @@
+#pragma GCC diagnostic pop
+#else
+#include "rascal/external/json.hpp"
+#endif
```
its not 'bad', g++-9 now complains about a no-problem in this header. The alternative is to change the whole CMakeList.txt architecture to link it with the -issystem flag which would supress all warnings. I discussed this with @Luthaf and he agrees with me that the best of the possible solutions is this one.

mastricker replied:
Ok, granted :)

felixmusil commented:
> Okay, yeah, this is way too big to review and merge all in one go.

I really don't feel it is too large. This PR is about adding 2 features (sparse training and some io) which corresponds to the addition of 2 header files, a few small modifications here and there, appending 20 lines to some files to enable the json conversion, bindings of the new class, quite a bit of new python code. Maybe I have not been clear enough on the structure of the changes / in code documentation of these so please point out to me what is missing so I can answer/improve these.

These 2 features could be split but since (1) they are very orthogonal (changes don't really overlap), (2) that we have discussed at length these features and how it should be implemented over several meetings and (3) that they are the last stone to the SOAP training milestone that is 3 months behind schedule, I thought we could also review these 2 together.


> As @Luthaf mentioned, we can use Git to squash the commits that have uninformative messages and then tease apart the commits into separate branches. Both can be done with interactive rebases; perhaps it would go faster if you sat together with a reviewer/git expert and split up the changes accordingly.

I am really not confortable at doing this since I don't think it is as easy as you make it sound. If you think it has to be done then let's do that together.


> This artificial distinction is resulting in lots of duplication of code and functionality (and, as @mastricker mentioned, code in the wrong place - why is pseudo_points.hh in models, when lots of its functionality in fact resembles property_block_sparse.hh? This is not just code duplication, it's non-local duplication).

If I understand well you would rather have `pseudo_points.hh` (which is a feature container) in structure_manager folder than in the models folder ? I don't think it makes more sense as it has nothing to do with structure_managers while it is key to build a sparse kernel model. Let's discuss that next meeting. The only reason `property_block_sparse.hh` is in the tructure_manager folder is because it is tied to structure_managers.


> In fact, I think it's touching on a deeper issue, which is that properties should be agnostic to whether or not they're actually attached to a structure or not. There should be a way to get a Property[BlockSparse] that was computed on a structure (or set of structures), but no longer depends on that structure (even if it's updated; maybe we have to do this via copying, maybe not). 

We already discussed this point a while ago in a meeting and we decided that there should be some data container tied to structure managers and if it does not make sense to tie features with structure (the case of pseudo points) we would make different data containers. That is what I implemented so if you want to discuss it again this PR is not right place I think.

> everything is just a dot product.

tying the data container to structure managers is exactly to simplify the implementation of dot products between these containers. The implementation of said dot products without the link to the structure manager would be quite different so I don't really see the code overlap ?

> Finally, if this class is going to stick around anyway: I really, really don't like the name "pseudo-points". I don't care if it's standard terminology for KRR or whatever; the "pseudo" seems to imply that these are on a lower conceptual footing than the 'real' points (yes, those are scare quotes) computed on all structures, which is not the case. In fact, we should be emphasizing that they're effectively the basis functions of the fit; something like "BasisPoints" or "SupportPoints" seems much more fitting.

Well you don't care but I think that we should try as much as possible to stick to the domain specific terminology to avoid confusion. This class is about sparse kernel methods where they call these either pseudo or sparse points. I agree this might be an unfortunate jargon but if someone wants to understand what we are doing, he will probably go and look at an original paper (like this one Quiñonero-Candela, J. & Rasmussen, C. E. (2005). J. Mach. Learn. Res. 6, 1939–1959.) and if the terminology is different then it becomes much harder to make the link between what we have implemented and the method article.
Your interpretation of 'pseudo' is one possible, I think the orignial author came up with it to make a clear distinction with 'proper' samples since these pseudo points could be artificial and they don't have an associated property.

I emphasise in the documentation that these are effectively basis points. Is it not clear enough ?

max-veit commented:
> Thanks for you comments @max-veit. Are you not going to review this PR ? or did I misunderstand ?

Not if it means reviewing all these changes in one PR. I don't have the time or mental RAM to review all of these changes at once. Separate PRs would be much easier.

> > Okay, yeah, this is way too big to review and merge all in one go.
> 
> I really don't feel it is too large.

I strongly disagree. Our PRs have been getting way too large for quite a while, so maybe this one feels "not too bad" in comparison. But it's way too many entangled features.

> This PR is about adding 2 features (sparse training and some io) which corresponds to the addition of 2 header files, a few small modifications here and there, appending 20 lines to some files to enable the json conversion, bindings of the new class, quite a bit of new python code about the io and the sparse kernel model (but this is quite separated in the code).

Exactly. That's too many changes in one go. The github interface shows three code files where the diffs are too large to show by default. If you ask me, just one of those large diffs is too much.

> Maybe I have not been clear enough on the structure of the changes / in code documentation of these so please point out to me what is missing so I can answer/improve these.

It's not so much a question of documentation or structure. If the features can be separated, then separate them and submit them as separate PRs. And especially as we already have so many entangled features that really should be orthogonal, I think it would be a good idea to start reinforcing the practice of keeping PRs, and therefore features, smaller and less interdependent.

> These 2 features could be split but since (1) they are very orthogonal (changes don't really overlap), (2) that we have discussed at length these features and how it should be implemented over several meetings and (3) that they are the last stone to the SOAP training milestone that is 3 months behind schedule, I thought we could also review these 2 together.

These are all good arguments for separate PRs. As for point (2), I don't remember the discussion on the pseudopoints, and many of the code and structural changes would have been better served by opening draft PRs earlier so we can have the discussion over actual code before everything becomes integrated. And as for point (3): Yes, I agree we're behind schedule, but that's not a good reason to rush the last few "keystone" features and end up with a solution that costs us lots of time and incurs further delays later on.

> > As @Luthaf mentioned, we can use Git to squash the commits that have uninformative messages and then tease apart the commits into separate branches [...]
> 
> I am really not confortable at doing this since I don't think it is as easy as you make it sound. If you think it has to be done then let's do that together.

That's exactly what I'm proposing. I'm away next week, so maybe @Luthaf can help?

> > This artificial distinction is resulting in lots of duplication of code and functionality (and, as @mastricker mentioned, code in the wrong place - why is pseudo_points.hh in models, when lots of its functionality in fact resembles property_block_sparse.hh? This is not just code duplication, it's non-local duplication).
> 
> If I understand well you would rather have `pseudo_points.hh` (which is a feature container) in structure_manager folder than in the models folder ? I don't think it makes more sense as it has nothing to do with structure_managers while it is key to build a sparse kernel model. Let's discuss that next meeting. The only reason `property_block_sparse.hh` is in the tructure_manager folder is because it is tied to structure_managers.

Agreed, let's discuss this next meeting, because now I have the feeling that both the feature container (which is actually also a much better name than `pseudo_points`) and the `property*.hh` data structures -- which are really just feature containers than can be indexed like structure managers -- belong in their own folder now. As for `property*.hh` being tied to structure managers: I would rather say "entangled". The interface between the "attached" properties and the structure manager is too deep and ill-defined; when you think about it, it really just boils down to mapping cluster indices (from the structure) to integer indices into the feature array. This lack of a well-defined separation and interface is, I'm guessing, what led to the duplication of feature-manager functionality into `pseudo_points.hh` in the first place.

> > In fact, I think it's touching on a deeper issue [...]
> 
> We already discussed this point a while ago in a meeting and we decided that there should be some data container tied to structure managers and if it does not make sense to tie features with structure (the case of pseudo points) we would make different data containers.

Are you referring to the discussion last April where we decided to make the "representation calculator" change? In any case, I think it's worth having that discussion again, because I'm no longer sure it makes sense to have representations _tied_ to structure managers in the way they are now. If anything, the present `property*` classes could probably just be reduced to interfaces from a structure manager to a feature container.

> That is what I implemented so if you want to discuss it again this PR is not right place I think.

We can discuss it in the next meeting (well, maybe the one after that, unless you want to skype me in from the Aiguille du Midi), but it should absolutely be done before this PR is merged. It'll be much harder to change these design decisions once they become part of the master branch and other features start building on them.

> > everything is just a dot product.
> 
> tying the data container to structure managers is exactly to simplify the implementation of dot products between these containers. The implementation of said dot products without the link to the structure manager would be quite different so I don't really see the code overlap ?

How so? What exactly in the dot product depends on the existence of a structure manager? It's either atom-wise (or cluster-wise), in which case we retain the respective atom/cluster indices which, in the property, can be collapsed to simple ints. Or, if it's structure-wise, we just sum over the clusters. What part of that depends on the existence of a structure manager?

> > Finally, if this class is going to stick around anyway: I really, really don't like the name "pseudo-points". I don't care if it's standard terminology for KRR or whatever; the "pseudo" seems to imply that these are on a lower conceptual footing than the 'real' points (yes, those are scare quotes) computed on all structures, which is not the case. In fact, we should be emphasizing that they're effectively the basis functions of the fit; something like "BasisPoints" or "SupportPoints" seems much more fitting.
> 
> Well you don't care but I think that we should try as much as possible to stick to the domain specific terminology to avoid confusion. This class is about sparse kernel methods where they call these either pseudo or sparse points. I agree this might be an unfortunate jargon but if someone wants to understand what we are doing, he will probably go and look at an original paper (like this one Quiñonero-Candela, J. & Rasmussen, C. E. (2005). J. Mach. Learn. Res. 6, 1939–1959.) and if the terminology is different then it becomes much harder to make the link between what we have implemented and the method article.

That's probably not the kind of paper most users will go to if they want to understand what we're doing. Remember, most users of this code will likely be chemists, physicists, and materials scientists, not computer scientists. Using terminology that comes from pure machine-learning research is likely to mislead such users, especially if they're more familiar with the GAP method (which calls these "sparse points", IMHO much closer to "basis points" than "pseudo-points").

> Your interpretation of 'pseudo' is one possible, I think the orignial author came up with it to make a clear distinction with 'proper' samples since these pseudo points could be artificial and they don't have an associated property.

It's still misleading and makes it sound like these points aren't as important in the regression as the points that correspond directly to data samples ("proper" points, if you must, though that's just as misleading as "pseudo"). We don't have to keep bad terminology if it's from a different field.

> I emphasise in the documentation that these are effectively basis points. Is it not clear enough ?

It would be better to choose a name that corresponds more closely to our usage (`feature container` or `basis points`), then point out in the documentation that these are effectively the same as the things called "pseudo points" in that paper.

This is a lot of discussion for one PR, but it's a lot of important changes. These definitely warrant more discussion as a group, so let's make sure we have a group meeting as soon as we're all in the same country again.

felixmusil commented:
> Not if it means reviewing all these changes in one PR. I don't have the time or mental RAM to review all of these changes at once. Separate PRs would be much easier.

I see so let me stress again that these features will not be merged in master before March then.

> Exactly. That's too many changes in one go. The github interface shows three code files where the diffs are too large to show by default. If you ask me, just one of those large diffs is too much.

What happens when ONE feature requires more changes than the default display of github ? 
Here I think you are refering to the 2 header files that have been added. Should I split these files into *_impl_1.hh, *_impl_2.hh untill all files are displayed fully in the github interface ?

>  (2), I don't remember the discussion on the pseudopoints, and many of the code and structural changes would have been better served by opening draft PRs earlier so we can have the discussion over actual code before everything becomes integrated.

Well this might have been meetings that you skiped and/or tuned out from but these discussions took place. So since these were already decided upon I don't really see the point of a draft PR unless our development meetings are meaningless ?

> The interface between the "attached" properties and the structure manager is too deep and ill-defined; [...]
In any case, I think it's worth having that discussion again, because I'm no longer sure it makes sense to have representations tied to structure managers in the way they are now. If anything, the present property* classes could probably just be reduced to interfaces from a structure manager to a feature container.

How is it ill defined ? `property*.hh` are templated with their corresponding manager and hold a pointer to it. 
What you are challenging is, I think, at the core of the library at the moment. I am not saying we should discard your ideas, but since it would correspond to very deep changes into the code, please find some motivations that would warant a few months of additional development so that we can start the discussion.

> How so? What exactly in the dot product depends on the existence of a structure manager? It's either atom-wise (or cluster-wise), in which case we retain the respective atom/cluster indices which, in the property, can be collapsed to simple ints. Or, if it's structure-wise, we just sum over the clusters. What part of that depends on the existence of a structure manager?

You are right when it comes down to dot products between representation but when you want to do the dot product between representation gradients and representation then you need to iterate of the neighbors (for sperical_invariants).

> That's probably not the kind of paper most users will go to if they want to understand what we're doing. Remember, most users of this code will likely be chemists, physicists, and materials scientists, not computer scientists. Using terminology that comes from pure machine-learning research is likely to mislead such users, especially if they're more familiar with the GAP method (which calls these "sparse points", IMHO much closer to "basis points" than "pseudo-points").

Ok should I move to "sparse points" throughout the code ? which is the alternative jargon term for  this object.

> This is a lot of discussion for one PR, but it's a lot of important changes. These definitely warrant more discussion as a group, so let's make sure we have a group meeting as soon as we're all in the same country again.

Agreed

mastricker commented:
Not commenting further now, just want to encourage both of you to put more details in the meeting minutes I already added or add new points so we can have a start right away in the next (possible) meeting.

felixmusil commented:
has been merged in master with #262 and #267.
